asapdiscovery-modeling/asapdiscovery/modeling/scripts/slice_fragalysis_structures_with_mdtraj.py

```diff
@@ -0,0 +1,131 @@
+"""
+This script slices the structures from the fragalysis database into the active site and full protein PDB files,
+as well as the corresponding numpy arrays of the coordinates of the atoms in the active site and full protein.
+"""
+import argparse
+import logging
+import multiprocessing as mp
+from datetime import datetime
+from pathlib import Path
+
+import mdtraj as md
+import numpy as np
+from asapdiscovery.data.utils import check_filelist_has_elements
+
+
+################################################################################
+def get_args():
+    parser = argparse.ArgumentParser(description="")
+
+    # Input arguments
+    parser.add_argument(
+        "-d",
+        "--structure_dir",
+        required=True,
+        help="Path to structure directory.",
+    )
+    parser.add_argument(
+        "-o",
+        "--output_dir",
+        required=True,
+        help="Path to output directory.",
+    )
+
+    parser.add_argument(
+        "-l",
+        "--log_file",
+        type=str,
+        default="save_combined_frag_structures.log",
+        help="Log file",
+    )
+
+    # Performance arguments
+    parser.add_argument(
+        "-n",
+        "--num_cores",
+        type=int,
+        default=1,
+        help="Number of concurrent processes to run.",
+    )
+
+    return parser.parse_args()
+
+
+def analyze_mp(fn, out_dir):
+    full_protein_selection = "not element H and (chainid 0 or chainid 2)"
+    active_site_string = "not element H and (chainid 0 or chainid 2) and (residue 140 to 145 or residue 163 or residue 172 or residue 25 to 27 or residue 41 or residue 49 or residue 54 or residue 165 to 168 or residue 189 to 192)"
+    output_name = fn.stem
+
+    # Prepare logger
+    handler = logging.FileHandler(out_dir / f"{output_name}-log.txt", mode="w")
+    prep_logger = logging.getLogger(output_name)
+    prep_logger.setLevel(logging.INFO)
+    prep_logger.addHandler(handler)
+    prep_logger.info(datetime.isoformat(datetime.now()))
+
+    # Check if outputs exists
+    def check_output():
+        for fn_suffix in ["_acive_site", "_full_protein"]:
+            for extension in [".pdb", ".npy"]:
+                if not (out_dir / f"{output_name}{fn_suffix}{extension}").exists():
+                    return False
+        return True
+
+    if check_output():
+        prep_logger.info("Output already exists!")
+        return True
+
+    # Load pdb
+    prep_logger.info(f"Loading {fn}")
+    pdb = md.load(fn)
+
+    # Slice pdb on active site
+    active_site_idx = pdb.topology.select(active_site_string)
+    active_site = pdb.atom_slice(active_site_idx)
+
+    # Slice pdb on full protein
+    full_protein_idx = pdb.topology.select(full_protein_selection)
+    full_protein = pdb.atom_slice(full_protein_idx)
+
+    # Save outputs of sliced structures to arrays of the indexes and pdb files
+    prep_logger.info(f"Saving idx arrays to {out_dir}")
+    np.save(out_dir / f"{output_name}_active_site.npy", active_site_idx)
```

Comment:
I would avoid passing around arrays of indices if it can be avoided as it can be error prone, better to use a well formed structure file of some kind and reconstruct them if that makes sense.

Reply:
this is more for keeping track of the indices that were used than for actually using them later